### PR TITLE
[Form] [ChoiceType] Prefer placeholder to empty_value

### PR DESCRIPTION
**Pull request by boite**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #16885 |
| License | MIT |
| Doc PR | - |

Prefer an explicitly set `placeholder` option (i.e. `false` or a non-empty
string) to an `empty_value` option when both are set.

The fix is to change the behaviour in the placeholder normalizer in
ChoiceType::configureOptions so that the value of the `empty_value` option is
used for placeholder only when the value of `placeholder` is null or an empty
string.

### CODE REVIEW

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php [hunk 1]
```diff
@@ -1726,6 +1726,140 @@ public function getOptionsWithPlaceholder()
...
+            'A placeholder is not used if it is explicitly set to false' => array(false, false, false, false, null, null),
+            'A placeholder is not used if it is explicitly set to false' => array(false, false, false, false, '', null),
+            'A placeholder is not used if it is explicitly set to false' => array(false, false, false, false, 'bar', null),
+            'A placeholder is not used if empty_value is set to false [maintains BC]' => array(false, false, false, null, false, null),
+            'An unset empty_value is automaticaly made an empty string in a non-required field (but null is expected here) [maintains BC]' => array(false, false, false, null, null, ''),
+            'An empty string empty_value is used if placeholder is not set [maintains BC]' => array(false, false, false, null, '', ''),
+            'A non-empty string empty_value is used if placeholder is not set [maintains BC]' => array(false, false, false, null, 'bar', 'bar'),
+            'A placeholder is not used if it is an empty string and empty_value is set to false [mainatins BC]' => array(false, false, false, '', false, null),
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php]: typo "maintains"

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php [hunk 2]
```diff
@@ -1726,6 +1726,140 @@ public function getOptionsWithPlaceholder()
...
+            'A placeholder is not used if it is explicitly set to false' => array(false, false, true, false, false, null),
+            'A placeholder is not used if it is explicitly set to false' => array(false, false, true, false, null, null),
+            'A placeholder is not used if it is explicitly set to false' => array(false, false, true, false, '', null),
+            'A placeholder is not used if it is explicitly set to false' => array(false, false, true, false, 'bar', null),
+            'A placeholder is not used if empty_value is set to false [maintains BC]' => array(false, false, true, null, false, null),
+            'A placeholder is not used if empty_value is not set [maintains BC]' => array(false, false, true, null, null, null),
+            'An empty string empty_value is used if placeholder is not set [maintains BC]' => array(false, false, true, null, '', ''),
+            'A non-empty string empty_value is used if placeholder is not set [maintains BC]' => array(false, false, true, null, 'bar', 'bar'),
+            'A placeholder is not used if it is an empty string and empty_value is set to false [mainatins BC]' => array(false, false, true, '', false, null),
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php]: same here

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php [hunk 3]
```diff
@@ -1726,6 +1726,140 @@ public function getOptionsWithPlaceholder()
...
+            'A placeholder is not used if it is explicitly set to false' => array(false, true, false, false, null, null),
+            'A placeholder is not used if it is explicitly set to false' => array(false, true, false, false, '', null),
+            'A placeholder is not used if it is explicitly set to false' => array(false, true, false, false, 'bar', null),
+            'A placeholder is not used if empty_value is set to false [maintains BC]' => array(false, true, false, null, false, null),
+            'An unset empty_value is automaticaly made an empty string in a non-required field (but null is expected here) [maintains BC]' => array(false, true, false, null, null, null),
+            'An empty string empty_value is converted to "None" in an expanded single choice field [maintains BC]' => array(false, true, false, null, '', 'None'),
+            'A non-empty string empty_value is used if placeholder is not set [maintains BC]' => array(false, true, false, null, 'bar', 'bar'),
+            'A placeholder is not used if it is an empty string and empty_value is set to false [mainatins BC]' => array(false, true, false, '', false, null),
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php]: same here

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php [hunk 4]
```diff
@@ -1726,6 +1726,140 @@ public function getOptionsWithPlaceholder()
...
+            'A placeholder is not used if it is explicitly set to false' => array(false, true, true, false, null, null),
+            'A placeholder is not used if it is explicitly set to false' => array(false, true, true, false, '', null),
+            'A placeholder is not used if it is explicitly set to false' => array(false, true, true, false, 'bar', null),
+            'A placeholder is not used if empty_value is set to false [maintains BC]' => array(false, true, true, null, false, null),
+            'A placeholder is not used if empty_value is not set [maintains BC]' => array(false, true, true, null, null, null),
+            'An empty string empty_value is converted to "None" in an expanded single choice field [maintains BC]' => array(false, true, true, null, '', 'None'),
+            'A non-empty string empty_value is used if placeholder is not set [maintains BC]' => array(false, true, true, null, 'bar', 'bar'),
+            'A placeholder is not used if it is an empty string and empty_value is set to false [mainatins BC]' => array(false, true, true, '', false, null),
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php]: same here

**Comment by HeahDude** [comment]
:+1: except your last test being very difficult to read.

Perhaps you could organise lines by BC first, then by expanded or multiple.
Rephrasing might help like :
"If placeholder is null", "If placeholder is false and empty_value is empty" ...

**Comment by xabbuh** [comment]
This fix looks good, but we really need so many test fixtures? Can we reduce them to the bare minimum?

**Comment by boite** [comment]
@xabbuh the (admittedly extensive) test data were invaluable in that they revealed when the fix was finally correct after a number of failed attempts. Surely, one datum for each possible permutation *is* the bare minimum.

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php [hunk 5]
```diff
@@ -1726,6 +1726,140 @@ public function getOptionsWithPlaceholder()
...
+     * @group legacy
+     */
+    public function testPlaceholderOptionWithEmptyValueOption(
+        $multiple,
+        $expanded,
+        $required,
+        $placeholder,
+        $emptyValue,
+        $viewValue
+    ) {
```
Comment (fabpot) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php]: should be moved on one line

**Comment by fabpot** [comment]
:+1:

**Comment by webmozart** [comment]
:+1:

**Comment by fabpot** [comment]
Thank you @boite.

File: src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php [hunk 6]
```diff
@@ -1726,6 +1726,134 @@ public function getOptionsWithPlaceholder()
...
+
+    public function getOptionsWithPlaceholderAndEmptyValue()
+    {
+        return array(
+            // single non-expanded, not required
+            'A placeholder is not used if it is explicitly set to false' => array(false, false, false, false, false, null),
+            'A placeholder is not used if it is explicitly set to false' => array(false, false, false, false, null, null),
+            'A placeholder is not used if it is explicitly set to false' => array(false, false, false, false, '', null),
+            'A placeholder is not used if it is explicitly set to false' => array(false, false, false, false, 'bar', null),
+            'A placeholder is not used if empty_value is set to false [maintains BC]' => array(false, false, false, null, false, null),
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php]: All those tests may not run because some keys are overridden...

Comment (fabpot) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php]: indeed

Comment (HeahDude) [inline src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php]: I'm on it

**Comment by HeahDude** [comment]
Here's the actual report :
```
There were 8 failures:

1) Symfony\Component\Form\Tests\Extension\Core\Type\ChoiceTypeTest::testPlaceholderOptionWithEmptyValueOption with data set "An empty string empty_value is converted to "None" in an expanded single choice field when expanded and required [maintains BC]" (false, true, true, null, '', 'None')
Failed asserting that null is identical to 'None'.

/Users/Heah/Sites/dev/symfony/src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php:1913

2) Symfony\Component\Form\Tests\Extension\Core\Type\ChoiceTypeTest::testPlaceholderOptionWithEmptyValueOption with data set "A non-empty string empty_value is used if placeholder is not set when expanded and required [maintains BC]" (false, true, true, null, 'bar', 'bar')
Failed asserting that null is identical to 'bar'.

/Users/Heah/Sites/dev/symfony/src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php:1913

3) Symfony\Component\Form\Tests\Extension\Core\Type\ChoiceTypeTest::testPlaceholderOptionWithEmptyValueOption with data set "An empty string empty_value is converted to "None" in an expanded single choice field when required [maintains BC]" (false, true, true, '', '', 'None')
Failed asserting that null is identical to 'None'.

/Users/Heah/Sites/dev/symfony/src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php:1913

4) Symfony\Component\Form\Tests\Extension\Core\Type\ChoiceTypeTest::testPlaceholderOptionWithEmptyValueOption with data set "A non-empty string empty_value is used if placeholder is an empty string when expanded and required [maintains BC]" (false, true, true, '', 'bar', 'bar')
Failed asserting that null is identical to 'bar'.

/Users/Heah/Sites/dev/symfony/src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php:1913

5) Symfony\Component\Form\Tests\Extension\Core\Type\ChoiceTypeTest::testPlaceholderOptionWithEmptyValueOption with data set "A non-empty string placeholder takes precedence over an empty_value set to false when expanded and required" (false, true, true, 'foo', false, 'foo')
Failed asserting that null is identical to 'foo'.

/Users/Heah/Sites/dev/symfony/src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php:1913

6) Symfony\Component\Form\Tests\Extension\Core\Type\ChoiceTypeTest::testPlaceholderOptionWithEmptyValueOption with data set "A non-empty string placeholder takes precedence over a not set empty_value when expanded and required" (false, true, true, 'foo', null, 'foo')
Failed asserting that null is identical to 'foo'.

/Users/Heah/Sites/dev/symfony/src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php:1913

7) Symfony\Component\Form\Tests\Extension\Core\Type\ChoiceTypeTest::testPlaceholderOptionWithEmptyValueOption with data set "A non-empty string placeholder takes precedence over an empty string empty_value when expanded and required" (false, true, true, 'foo', '', 'foo')
Failed asserting that null is identical to 'foo'.

/Users/Heah/Sites/dev/symfony/src/Symfony/Component/Form/Tests/Extension/Core/Type/ChoiceTypeTest.php:1913

8) Symfony\Component\Form\Tests\Extension\Core\Type\ChoiceTypeTest::testPlaceholderOptionWithEmptyValueOption with data set "A non-empty string placeholder takes precedence over a non-empty string empty_value when expanded and required" (false, true, true, 'foo', 'bar', 'foo')
Failed asserting that null is identical to 'foo'.
```

**Comment by boite** [comment]
Damn it! Will you let me sort this out?  I'll do a better job of naming the data keys for readability (and uniqueness).

**Comment by HeahDude** [comment]
I fixed it, but I still can hold the PR. What do I do ?

**Comment by HeahDude** [comment]
@boite, see #17988, feel free to send a PR that better suits you and I'll close this one.

**Pull request closed without merging**